src/fileprovider.cs

```diff
@@ -0,0 +1,307 @@
+﻿﻿//
+// FileProvider C# bindings
+//
+// Authors:
+//	Alex Soto  <alexsoto@microsoft.com>
+//
+// Copyright 2017 Xamarin Inc. All rights reserved.
+//
+
+#if XAMCORE_2_0
+
+using System;
+using XamCore.ObjCRuntime;
+using XamCore.CoreGraphics;
+using XamCore.Foundation;
+
+namespace XamCore.FileProvider {
+
+	[iOS (11,0)]
+	[ErrorDomain ("NSFileProviderErrorDomain")]
+	[Native]
+	public enum NSFileProviderError : nint {
+		NotAuthenticated = -1000,
+		FilenameCollision = -1001,
+		SyncAnchorExpired = -1002,
+		PageExpired = SyncAnchorExpired,
+		InsufficientQuota = -1003,
+		ServerUnreachable = -1004,
+		NoSuchItem = -1005,
+	}
+
+	[iOS (11,0)]
+	enum NSFileProviderItemIdentifier {
+
+		[Field ("NSFileProviderRootContainerItemIdentifier")]
+		Root,
+
+		[Field ("NSFileProviderWorkingSetContainerItemIdentifier")]
+		WorkingSet,
+	}
+
+	[iOS (11,0)]
+	[Native]
+	[Flags]
+	public enum NSFileProviderItemCapabilities : nuint {
+		Reading = 1 << 0,
+		Writing = 1 << 1,
+		Reparenting = 1 << 2,
+		Renaming = 1 << 3,
+		Trashing = 1 << 4,
+		Deleting = 1 << 5,
+		AddingSubItems = Writing,
+		ContentEnumerating = Reading,
+		All = Reading | Writing | Reparenting | Renaming | Trashing | Deleting,
+	}
+
+	[iOS (11,0)]
+	[DisableDefaultCtor]
+	[BaseType (typeof (NSObject))]
+	interface NSFileProviderDomain {
+
+		[Export ("initWithIdentifier:displayName:pathRelativeToDocumentStorage:")]
+		IntPtr Constructor (string identifier, string displayName, string pathRelativeToDocumentStorage);
+
+		[Export ("identifier")]
+		string Identifier { get; }
+
+		[Export ("displayName")]
+		string DisplayName { get; }
+
+		[Export ("pathRelativeToDocumentStorage")]
+		string PathRelativeToDocumentStorage { get; }
+	}
+
+	interface INSFileProviderEnumerationObserver { }
+
+	[iOS (11,0)]
+	[Protocol]
+	interface NSFileProviderEnumerationObserver {
+
+		[Abstract]
+		[Export ("didEnumerateItems:")]
+		void DidEnumerateItems (INSFileProviderItem [] updatedItems);
+
+		[Abstract]
+		[Export ("finishEnumeratingUpToPage:")]
+		void FinishEnumerating ([NullAllowed] NSData nextPage);
```

Comment:
I think the parameter name should be closer to the ObjC selector, to avoid confusion (`NSData upToPage` doesn't sound that bad to me).